### ISSUE TITLE
Add Qt

### ISSUE THREAD

**Issue by ruslo** [ruslo/hunter#116]

**Comment by ruslo** [comment]
Android? iOS?

**Comment by pretyman** [comment]
What would be strategies for adding support for Qt? 
1. Download the platform dependent pre-compiled compressed file and just uncompress it into `<root-id>/Install`
2. Download the open source source compressed, compile it locally and install it to `<root-id>/Install`
3. Make it in a way where both of the above options will be possible

For iOS + Qt support, what would be best to figure out how the makefiles are generated?
1. Look into qmake's source code to see how the make files are generated for iOS
2. Reverse engineer the Makefiles generated by qmake when run -spec macx-ios-clang CONFIG+=iphoneos

**Comment by ruslo** [comment]
> 1. Download the platform dependent pre-compiled compressed file and just uncompress it into <root-id>/Install

I think no because it's not possible to apply patches in this case. Also it would be nice to have configuration options (like with/without OpenSSL, shared/static, ...)

> 2. Download the open source source compressed, compile it locally and install it to <root-id>/Install

Yes

> 3. Make it in a way where both of the above options will be possible

I don't think it's worth it to make only for one package. Downloading binaries should be implemented in more general way, see https://github.com/ruslo/hunter/issues/75.

> For iOS + Qt support, what would be best to figure out how the makefiles are generated?

We just need to follow official instructions and save all knowledge in build scheme.

**Comment by pretyman** [comment]
> We just need to follow official instructions and save all knowledge in build scheme.
Sorry for not being clear. I was not referring to Qt, but the iOS projects based on Qt, once at the moment is done by qmake -spec macx-ios-clang CONFIG+=iphoneos.

**Comment by ruslo** [comment]
> Sorry for not being clear. I was not referring to Qt, but the iOS projects based on Qt, once at the moment is done by qmake -spec macx-ios-clang CONFIG+=iphoneos.

I think Qt is CMake-friendly, i.e. to integrate Qt into CMake project we just need `find_package(Qt*)` ?). I haven't tried iOS variant of Qt yet.

**Comment by pretyman** [comment]
Qt for Desktop was indeed CMake-friendly and straight forward, but I haven't yet managed to get it to work for iOS. I will investigate this further later on.

I made a work in progress `project-qt` branch on my fork: https://github.com/pretyman/hunter/tree/project-qt

What is happening is that the `prefix` seems not respected and the installation is landing in `/tmp/Hunter.dst/...` as per:

```
install -m 644 -p /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qttranslations/translations/qtxmlpatterns_ko.qm /tmp/Hunter.dst/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Install/translations/
install -m 644 -p /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qttranslations/translations/qtxmlpatterns_ru.qm /tmp/Hunter.dst/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Install/translations/
install -m 644 -p /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qttranslations/translations/qtxmlpatterns_sk.qm /tmp/Hunter.dst/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Install/translations/
install -m 644 -p /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qttranslations/translations/qtxmlpatterns_uk.qm /tmp/Hunter.dst/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Install/translations/
cd /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source && /opt/local/bin/cmake -E touch /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Build/Qt-prefix/src/Qt-stamp/Debug-iphoneos/Qt-install
```

**Comment by ruslo** [comment]
> What is happening is that the prefix seems not respected

I see you're using:
```cmake
    CONFIGURE_COMMAND
    ${configure_command}
    ${configure_opts}
    "--prefix=@HUNTER_PACKAGE_INSTALL_PREFIX@"
```

But from [documentation](http://doc.qt.io/qt-5/configure-options.html) we need `-prefix`. So I guess it should be:
```cmake
    CONFIGURE_COMMAND
    ${configure_command}
    ${configure_opts}
    -prefix
    "@HUNTER_PACKAGE_INSTALL_PREFIX@"
```

?

**Comment by pretyman** [comment]
I also noticed that, but since Qt for desktop went smoothly, I thought it would be ok. I will make your suggested change and try again.

**Comment by ruslo** [comment]
@pretyman Hi Alexandre. I've pushed your changes to branch [project-qt](https://github.com/ruslo/hunter/tree/project-qt) since I want to experiment how it works on iOS. Please don't rebase your local branch since we may have tons of conflicts :) Also feel free to update/merge from it.

**Comment by ruslo** [comment]
Some links to test Qt for iOS:
* http://jwintz.me/blog/2013/03/08/qt-on-ios-examples/
* http://stackoverflow.com/questions/25042916/run-time-error-for-qt-application-on-ios-built-via-cmake
* https://github.com/jwintz/qt-ios-examples

**Comment by pretyman** [comment]
@ruslo I tried your suggested change of the -prefix yesterday, but still got the same result of the installation going to `/tmp/hunter.dst/...` however, у меня нето лицензии, я сделал изменения в исходный код в mkspecs и, наверное, поэтому не работал. If you have a proper setup, it should work.

**Comment by ruslo** [comment]
> If you have a proper setup, it should work

Qt is building all day today :) This in progress. I will let you know.

**Comment by ruslo** [comment]
Also:

`у меня нето лицензии` -> `у меня нет лицензии`

`я сделал изменения в исходный код в mkspecs` -> `я внёс изменения в исходный код в mkspecs`
or
`я сделал изменения в исходный код в mkspecs` -> `я сделал изменения в исходном коде в mkspecs`

Though I'm not sure what is mkspecs exactly, is it a directory or name of an option?

`поэтому не работал` ->  `поэтому не работает` (not working)
or
`поэтому не работал` ->  `поэтому не сработало` (haven't work)
or
`поэтому не работал` ->  `поэтому не работало` (didn't work)

**Comment by ruslo** [comment]
> Still in progress. I will let you know

@pretyman It works! Amazing :) I've pushed updates to branch `qt-project`. Thank you again!

**Comment by ruslo** [comment]
### TODO

`-skip` option can be used to exclude the module from the build. List of the modules can be taken from the list of submodules:

```
qt3d
qtactiveqt
qtandroidextras
qtbase
qtcanvas3d
qtconnectivity
qtdeclarative
qtdoc
qtdocgallery
qtenginio
qtfeedback
qtgraphicaleffects
qtimageformats
qtlocation
qtmacextras
qtmultimedia
qtpim
qtqa
qtquick1
qtquickcontrols
qtrepotools
qtscript
qtsensors
qtserialport
qtsvg
qtsystems
qttools
qttranslations
qtwayland
qtwebchannel
qtwebengine
qtwebkit
qtwebkit-examples
qtwebsockets
qtwinextras
qtx11extras
qtxmlpatterns
```

Links:
* [Submodules](http://code.qt.io/cgit/qt/qt5.git/tree/.gitmodules)
* [Configure options](http://doc.qt.io/qt-5/configure-options.html)

**Comment by pretyman** [comment]
Note that you cannot `-skip` the `qtbase` module

Sorry for not being complete, I was at lunch time. The reason I got the .dst problem was that я добавил `CODE_SIGN_IDENTITY=\"\" CODE_SIGNING_REQUIRED=NO` в архив `mkspecs/macx-ios-clang/features/default_post.prf` в строке 99:
```
xcodebuild = "xcodebuild CODE_SIGN_IDENTITY=\"\" CODE_SIGNING_REQUIRED=NO $$action -scheme $(TARGET) -sdk $$sdk -configuration $$title($$cfg)"
```
`mkspecs` is the directory which keeps the building specifications used by `qmake`

I did try the polly `build.py --toolchain ios-nocodesign` toolchain, but it seems to be failing at qxmlpatterns with the log message at the end. I'll just invest in a developer license from Apple when my app matures.

Thanks for the Russian cases, I jotted them down and will break them down. I've been learning for years, but still can't communicate properly.

Log of the failed build with polly toolchain `ios-nocodesign`, in case you have tried and got a different result:

```
codebuild build -scheme xmlpatternsvalidator -sdk iphoneos -configuration Release
Build settings from command line:
    SDKROOT = iphoneos8.1
Build settings from configuration file '/Users/discipline/polly/scripts/NoCodeSign.xcconfig':
    CODE_SIGNING_REQUIRED = NO 
Build settings from command line:
    SDKROOT = iphoneos8.1
Build settings from configuration file '/Users/discipline/polly/scripts/NoCodeSign.xcconfig':
    CODE_SIGNING_REQUIRED = NO 
2015-07-14 16:07:18.940 xcodebuild[34073:1007] +dataWithFirstBytes:1024 ofFile:"/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/.qmake.cache" failed, errno = 2
2015-07-14 16:07:18.940 xcodebuild[34076:1007] +dataWithFirstBytes:1024 ofFile:"/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/.qmake.cache" failed, errno = 2
=== BUILD AGGREGATE TARGET Qt Preprocess OF PROJECT xmlpatterns WITH CONFIGURATION Release ===
=== BUILD AGGREGATE TARGET Qt Preprocess OF PROJECT xmlpatternsvalidator WITH CONFIGURATION Release ===
Check dependencies
Check dependencies
Write auxiliary files
/bin/mkdir -p /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator/.obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build
write-file /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator/.obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build/Script-5E618435888B9D49F8540165.sh
chmod 0755 /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator/.obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build/Script-5E618435888B9D49F8540165.sh
write-file /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator/.obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build/Script-7EF0942E79C014DCEC8976BC.sh
chmod 0755 /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator/.obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build/Script-7EF0942E79C014DCEC8976BC.sh
Write auxiliary files
/bin/mkdir -p /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns/.obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build
write-file /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns/.obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build/Script-7EF0942E79C014DCEC8976BC.sh
chmod 0755 /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns/.obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build/Script-7EF0942E79C014DCEC8976BC.sh
write-file /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns/.obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build/Script-5E618435888B9D49F8540165.sh
chmod 0755 /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns/.obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build/Script-5E618435888B9D49F8540165.sh
PhaseScriptExecution Qt\ Qmake .obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build/Script-5E618435888B9D49F8540165.sh
    cd /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns
    /bin/sh -c \"/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns/.obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build/Script-5E618435888B9D49F8540165.sh\"
PhaseScriptExecution Qt\ Qmake .obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build/Script-5E618435888B9D49F8540165.sh
    cd /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator
    /bin/sh -c \"/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator/.obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build/Script-5E618435888B9D49F8540165.sh\"
make[5]: warning: jobserver unavailable: using -j1.  Add `+' to parent make rule.
make[5]: warning: jobserver unavailable: using -j1.  Add `+' to parent make rule.
make[5]: `xmlpatternsvalidator.xcodeproj/project.pbxproj' is up to date.
make[5]: `xmlpatterns.xcodeproj/project.pbxproj' is up to date.
PhaseScriptExecution Qt\ Preprocessors .obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build/Script-7EF0942E79C014DCEC8976BC.sh
    cd /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator
    /bin/sh -c \"/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatternsvalidator/.obj/xmlpatternsvalidator.build/Release-iphoneos/Qt\ Preprocess.build/Script-7EF0942E79C014DCEC8976BC.sh\"
PhaseScriptExecution Qt\ Preprocessors .obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build/Script-7EF0942E79C014DCEC8976BC.sh
    cd /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns
    /bin/sh -c \"/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Source/qtxmlpatterns/tools/xmlpatterns/.obj/xmlpatterns.build/Release-iphoneos/Qt\ Preprocess.build/Script-7EF0942E79C014DCEC8976BC.sh\"
make[5]: warning: jobserver unavailable: using -j1.  Add `+' to parent make rule.
make[5]: Nothing to be done for `preprocess'.
make[5]: warning: jobserver unavailable: using -j1.  Add `+' to parent make rule.
=== BUILD TARGET xmlpatternsvalidator OF PROJECT xmlpatternsvalidator WITH CONFIGURATION Release ===
Check dependencies
make[5]: Nothing to be done for `preprocess'.
=== BUILD TARGET xmlpatterns OF PROJECT xmlpatterns WITH CONFIGURATION Release ===
Check dependencies
Code Sign error: No code signing identities found: No valid signing identities (i.e. certificate and private key pair) matching the team ID “(null)” were found.
Code Sign error: No code signing identities found: No valid signing identities (i.e. certificate and private key pair) matching the team ID “(null)” were found.
** BUILD FAILED **
The following build commands failed:
        Check dependencies
(1 failure)
** BUILD FAILED **
The following build commands failed:
        Check dependencies
(1 failure)
make[4]: *** [iphoneos-release] Error 65
make[3]: *** [iphoneos] Error 2
make[2]: *** [sub-xmlpatterns-make_first] Error 2
make[2]: *** Waiting for unfinished jobs....
make[4]: *** [iphoneos-release] Error 65
make[3]: *** [iphoneos] Error 2
make[2]: *** [sub-xmlpatternsvalidator-make_first] Error 2
make[1]: *** [sub-tools-make_first] Error 2
make: *** [module-qtxmlpatterns-make_first] Error 2
make: *** [/Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt/Build/Qt-prefix/src/Qt-stamp/Debug-iphoneos/Qt-build] Error 2
Showing first 200 notices only
Showing first 200 warnings only

** BUILD FAILED **


The following build commands failed:
        PhaseScriptExecution CMake\ Rules Build/Hunter.build/Debug-iphoneos/Qt.build/Script-A81159A750F149B9B13D6B5F.sh
(1 failure)

[hunter ** FATAL ERROR **] Build step failed (dir: /Users/discipline/.hunter/_Base/xxxxxxx/d1a0cd4/34d19ed/Build/Qt
[hunter ** FATAL ERROR **] [Directory:/Users/discipline/.hunter/cmake/projects/Qt]

------------------------------ WIKI -------------------------------
    https://github.com/ruslo/hunter/wiki/error.external.build.failed
-------------------------------------------------------------------

CMake Error at /Users/discipline/.hunter/cmake/modules/hunter_wiki.cmake:12 (message):
Call Stack (most recent call first):
  /Users/discipline/.hunter/cmake/modules/hunter_fatal_error.cmake:20 (hunter_wiki)
  /Users/discipline/.hunter/cmake/modules/hunter_download.cmake:384 (hunter_fatal_error)
  /Users/discipline/.hunter/cmake/projects/Qt/hunter.cmake:27 (hunter_download)
  /Users/discipline/.hunter/cmake/modules/hunter_add_package.cmake:87 (include)
  CMakeLists.txt:12 (hunter_add_package)


-- Configuring incomplete, errors occurred!
See also "/Users/discipline/.hunter/examples/qt-widgets/_builds/ios-nocodesign/CMakeFiles/CMakeOutput.log".
Command exit with status "1": [/Users/discipline/.hunter/examples/qt-widgets]> "cmake" "-H." "-B/Users/discipline/.hunter/examples/qt-widgets/_builds/ios-nocodesign" "-GXcode" "-DCMAKE_TOOLCHAIN_FILE=/Users/discipline/polly/ios-nocodesign.cmake" "-DCMAKE_VERBOSE_MAKEFILE=ON" "-DPOLLY_STATUS_DEBUG=ON" "-DHUNTER_STATUS_DEBUG=ON"

Log: /Users/discipline/.hunter/examples/qt-widgets/_builds/ios-nocodesign/_3rdParty/polly/log.txt
*** FAILED ***
```

**Comment by ruslo** [comment]
>  It works!

Well it was about `libcxx` toolchain. With `ios` toolchain I have the same install issue :(

Looks like Qt use additional prefix `/tmp/Hunter.dst`. So my package installed to `/tmp/Hunter.dst/<root-id>/Install` instead of `/<root-id>/Install` just like in your case.

**Comment by ruslo** [comment]
> List of the modules can be taken from the list of submodules

`configure` script reports that this modules doesn't exist:
```
qtdocgallery
qtfeedback
qtpim
qtqa
qtrepotools
qtsystems
```

**Comment by ruslo** [comment]
> Some links to test Qt for iOS:
>  http://jwintz.me/blog/2013/03/08/qt-on-ios-examples/
> https://github.com/jwintz/qt-ios-examples

Adapted: https://github.com/forexample/qt-ios-examples

**Comment by ruslo** [comment]
@pretyman I've merged some updates to branch `develop` (release [v0.12.0](https://github.com/ruslo/hunter/releases/tag/v0.12.0)):
* iOS install work for me now. It was fixed accidentally actually :) I was working on [cache size](https://github.com/ruslo/hunter/commit/3b5997b4e16b20345a3bd6706a426dc589cb5052) optimization and create a CMake script instead of `make install`. Looks like some environment variable wreak havoc before.
* Qt splitted into the components (see [wiki](https://github.com/ruslo/hunter/wiki/pkg.qt))
* Qt now support [CONFIGURATION_TYPES](https://github.com/ruslo/hunter/commit/bcf6955176e8e4728ee2d844f884a12aef22648f). You can set `Release`, `Debug` or `Release;Debug` (release + debug works only on OSX (?))
* Qt is [cacheable](https://github.com/ruslo/hunter/commit/05650295899ab60197423efe808f3effd8947a87)
* Support [static/shared](https://github.com/ruslo/hunter/commit/51c3aa2de589573bad77798383c7fd937842fca8) variants. Note that default variant is static for Hunter. For Qt it means that you need to link some extra libraries. See [example](https://github.com/ruslo/hunter/blob/develop/examples/qt-widgets/CMakeLists.txt).

Components + configuration type + cache feature will save huge amount of build type. For instance qt-widgets example now builds ~ 15 minutes (qtbase/Release only, desktop) instead of 5+ hours (all components, Debug/Release). Note that default build variant is `Debug;Release`

iOS build have some extra peculiarities:
* simulator's libraries not found by `find_package` though they built and installed. Hence it make no sense to use universal iOS toolchain
* even if only one build type specified (like '-release') Qt5Core.cmake trying to find debug variant of library and failed. Hence only '-debug-and-release' (default) variant can be used
* for now Qt do configuration with iOS default settings (i.e. it ignores settings from toolchain and will build 8.2 target with Xcode 6.2 instead of 8.4 target with Xcode 6.4)
* architectures not controlled by toolchain, e.g. it always arm64 + armv7 for device
* @headupinclouds minimal build time because of this issues is about 1 hour (2 architectures x 2 variants)

**Comment by ruslo** [comment]
> List of the modules can be taken from the list of submodules

I've looked at `qtbase/configure` and discovered that component is just a directory with name `qt*`. So for now list of components determined automatically.

**Comment by pretyman** [comment]
@ruslo Wow, this is great news man! Well done!

> (release + debug works only on OSX (?))

Yep, as per `qtbase/configure` script:

```bash
if [ "$XPLATFORM_MAC" = "no" -a "$CFG_DEBUG_RELEASE" = "yes" ]; then
    echo
    echo "WARNING: -debug-and-release is not supported outside of Mac OS X."
    echo "Qt can be built in release mode with separate debug information, so"
    echo "-debug-and-release is not necessary anymore"
    echo
fi
```

> even if only one build type specified (like '-release') Qt5Core.cmake trying to find debug variant of library and failed. 

Seems that CMake support is still rough on its edges. I was trying to get an app which uses QtQuick + QtMultimedia compiling and running on iOS last night and was running into linking errors. I was compiling for `Release` and  `Qt5QmlConfig.cmake` was not including all the needed dependencies, it included automatically its plugins with the following line: 

```CMake
file(GLOB pluginTargets "${CMAKE_CURRENT_LIST_DIR}/Qt5Qml_*Plugin.cmake")
```

and the files in the directory are:
```shell
Qt5Qml_QTcpServerConnection.cmake
Qt5Qml_QtQuick2Plugin.cmake
```
Meaning `Qt5Qml_QTcpServerConnection.cmake`, which was no **Plugin** suffix, was not being picked up. Upon fixing this, I was able to get the app to compile, but not yet run. I see you did some changes to the main of qt-widgets example, so I will try the same, however my app works with QtCreator + qmake. 

Question: do you know how can I access the C++ (and C) compiling flags when using the iOS toolchain? They are not being written to `CMAKE_CXX_FLAGS` nor `CMAKE_C_FLAGS`. I was trying to see if `xcodebuild` had any option for that, but the furthest I got was `xcodebuild -showBuildSettings`, but it doesn't give me all compile flags. The aim of accessing this is to use the autotools based projects.

Thanks a lot for getting this work

**Comment by ruslo** [comment]
> even if only one build type specified (like '-release') Qt5Core.cmake trying to find debug variant of library and failed. Hence only '-debug-and-release' (default) variant can be used

It's not true, my bad. I was testing version with [bug](https://github.com/ruslo/hunter/commit/c2d3e4d5bcb70986e51cf8549f639d9e79c7b0f2).

@headupinclouds minimal time of build for iOS is ~30min (qtbase, Release, device only)

**Comment by ruslo** [comment]
> I was trying to get an app which uses QtQuick + QtMultimedia compiling and running on iOS last night and was running into linking errors

@pretyman Yes, some extra libraries needed. See [example](https://github.com/ruslo/hunter/blob/develop/examples/qt-widgets/CMakeLists.txt). Note that static build is default for Hunter, for dynamic Qt variant some of them not necessary.

> Meaning Qt5Qml_QTcpServerConnection.cmake, which was no Plugin suffix, was not being picked up

I guess if you're sure that this is an error you can open a bug: https://bugreports.qt.io

> Question: do you know how can I access the C++ (and C) compiling flags when using the iOS toolchain? They are not being written to CMAKE_CXX_FLAGS nor CMAKE_C_FLAGS.

They are added by Xcode itself, so yes, you can't use `CMAKE_{C|CXX}_FLAGS`. If you're using [Polly toolchain](https://github.com/ruslo/polly/blob/master/os/iphone.cmake) there are some help variables like `IOS_SDK_VERSION` or `IPHONESIMULATOR_ROOT`. Also you may take a look at [OpenSSL iOS](https://github.com/ruslo/hunter/blob/master/cmake/schemes/url_sha1_openssl_ios.cmake.in) scheme.

**Comment by ruslo** [comment]
> simulator's libraries not found by find_package though they built and installed. Hence it make no sense to use universal iOS toolchain

https://bugreports.qt.io/browse/QTBUG-47314

**Comment by ruslo** [comment]
From the second thought I can add extra install step which will fuse simulator and device libraries into one universal lib using `lipo` tool. This will fix both `find_package can't find Qt when only simulator libs built` and `can't run both device and simulator in one Xcode project`.

@headupinclouds @pretyman let me know if you're interested in simulator archs

**Comment by pretyman** [comment]
@ruslo Thanks for the heads with the OpenSSL's iOS build scheme, it was very useful. 

To tell you the truth, I am more interested in seeing Qt in the real device, although the simulator is useful, I haven't used it often myself. My process is to compile for the desktop with the device aspect ratio and when I am satisfied with the results, I try it on the device. But I do it mainly as a hobby and university, not real work, so "your milage may vary".

> https://bugreports.qt.io/browse/QTBUG-47314

I think that we might be a bit disappointed, if I remember well, the CMake scripts for Qt were written by a guy called Stephen Kelly, who used to work at KDAB, but left around a year ago. The CMake scripts are actually built at build time (maybe install? I am not 100% sure). They are created by `qtbase/mkspecs/features/create_cmake.prf` using `qtbase/mkspecs/features/cmake_functions.prf`. When I get around to reproduce their generation, I will post it here.

**Comment by ruslo** [comment]
I just want to memorize the fact that it's not a Hunter problem. Issues in Qt must be solved in Qt's bugtracker. If somebody want to add a note or fix it - do it there.

Also I think it's not a big deal. As I already mentioned I can apply a workaround in Hunter. Since we are controlling install process we can customize it. For example I've added extra plugin to aviod explicit linking of frameworks and other missing libraries/flags in user's code: https://github.com/ruslo/hunter/commit/2b2143f299a5ab5b071adc97d05321de38c63ec8 (I will test and add similar plugin for desktop today). Also this is how Boost imported-style libraries added (like `Boost::system` or `Boost::filesystem`): https://github.com/ruslo/hunter/commit/80f057704b68aed3931ef8a30f055273c553f62e

**Comment by ruslo** [comment]
> For example I've added extra plugin to aviod explicit linking of frameworks and other missing libraries/flags in user's code

https://bugreports.qt.io/browse/QTBUG-47336

**Comment by pretyman** [comment]
Sorry about discussing Qt's internals here.

Your approach of creating the `Qt5Widgets_HunterPlugin.cmake` was excellent,  this pattern of having cmake list files as `Qt5Module_*Plugin.cmake` makes much faster to test the fixes, which could later be pushed upstream, if that is to be the case.

**Comment by ruslo** [comment]
> I will test and add similar plugin for desktop today

Applied to branch develop: https://github.com/ruslo/hunter/commit/e6bd59c58f96b07d5c2b824843a5d971da73f4e1

(note that you must use CMake 3.1+ to use INTERFACE_SOURCES feature or add `${QT_ROOT}/src/static_qt_plugins.cpp` to the target manually)

Qt bug: https://bugreports.qt.io/browse/QTBUG-47349

**Comment by ruslo** [comment]
Android should work now. New HelloGL2 example: https://github.com/forexample/android-cmake/tree/master/05-qt-hellogl2

**Comment by pretyman** [comment]
У нас проблема building Qt components related to `qtdeclarative` separately the way we do components today: if a component has qml (declarative) bindings, for example `qtmultimedia`, they will only be installed and compiled if `qtdeclarative` is not skipped. If `qtdeclarative` is compiled by itself, it does not take into account `qtmultimedia`.

The components that generate qml bindings are (at least):
```
qt3d
qtcanvas3d
qtconnectivity
qtenginio
qtgraphicaleffects
qtlocation
qtmultimedia
qtsensors
qtwebchannel
qtwebkit
qtwebsockets
```
Do you have any idea what we could do to fix this? I tried looking around how the component build rules are done, but the design seems to be to compile them one by one. 
I think that providing a meta component `all` that doesn't skip anything could be used as a fallback for cases where we don't foresee these interdependencies.

**Comment by ruslo** [comment]
> Do you have any idea what we could do to fix this?

We can write explicit rules about each component. Quite the same happenning with `qtbase` - every component must not skip it by `-skip qtbase`. Otherwise nothing will be build at all. We can apply same logic here. We need to create list `depends_on_qtdeclarative_component` and check if the current component is in list. If so do not do the skip. Take a look at `scripts/qt-configure.cmake`. I will be testing it now.

**Comment by pretyman** [comment]
Понял. But I understand that we end up always building `qtdeclarative`, even when it might not be desired.

What I was trying to do and didn't succeed, was from a `hunter_add_package` line like `hunter_add_package(Qt COMPONENTS qtmultimedia qtdeclarative)`, I was trying to make all packages available in a list during `scripts/qt-configure.cmake` time, to check if `qtdeclarative` is present, if so to not skip it.  But the full package list information is only available to `@HUNTER_SELF@/cmake/modules/hunter_add_package.cmake` and not carried on.

I am not sure that was a trying to do was a good idea from the caching point of view, if someone wanted `qtmultimedia`  without `qtdeclarative` for example. The way you proposed, of always building `qtdeclarative`, we avoid any complexity on the cache side.

**Comment by ruslo** [comment]
> for example qtmultimedia

Works for me actually:
```cmake
hunter_add_package(Qt COMPONENTS qtmultimedia)
find_package(Qt5Multimedia REQUIRED)
```

can you please add an example with error

**Comment by ruslo** [comment]
> But I understand that we end up always building qtdeclarative, even when it might not be desired.

No, same as `qtbase`. Qt use special directory for building which doesn't removed after build done (unlike the regular packages). So if you build one component `qt3d` + extra `qtdeclarative`, then `qtcanvas3d` + `qtdeclarative` - at the end `qtdeclarative` will be built once.

**Comment by pretyman** [comment]
Sorry for not being clear. `qtmultimedia` does compile by iteself, yes, however when compiling by iteself, it does not generate its QML bindings under the directory `<root-id>/Prefix/qml/QtMultimedia`, when both `qtdeclarative` and `qtmultimedia` are compiled together, the directory is created, and populated with the following files inside:
```
Video.qml
libdeclarative_multimedia.prl
libdeclarative_multimedia_debug.prl
qmldir
libdeclarative_multimedia.a
libdeclarative_multimedia_debug.a
plugins.qmltypes
```

**Comment by ruslo** [comment]
> and populated with the following files inside

thanks, see no such files indeed

**Comment by ruslo** [comment]
I've pushed new branch `qt-dev` with updates. Looks like qml files installed too. Also now `7z` archive used for Qt since it's smaller and have `configure.exe` for windows. I've decide to merge this two updates because in this case we don't need to make a change of PACKAGE_INTERNALS_DEPS_ID. Want to test it more before merging to `develop`.

**Comment by pretyman** [comment]
Thanks, I see the changes. However I didn't manage to download the `7z` archive, for reasons I don't know. I tried many times (10?) however it always fail to download

```
-- [download 38% complete]
-- [download 39% complete]
-- [download 40% complete]
-- [download 41% complete]
-- [download 42% complete]
CMake Error at Build/Qt/__qtbase/Build/Qt-qtbase-prefix/src/Qt-qtbase-stamp/download-Qt-qtbase.cmake:27 (message):
  error: downloading
  'http://download.qt.io/official_releases/qt/5.5/5.5.0/single/qt-everywhere-opensource-src-5.5.0.7z'
  failed

    status_code: 18
    status_string: "Transferred a partial file"
    log:   Trying 77.86.229.90...

  Connected to download.qt.io (77.86.229.90) port 80 (#0)

  GET
  /official_releases/qt/5.5/5.5.0/single/qt-everywhere-opensource-src-5.5.0.7z
  HTTP/1.1

  Host: download.qt.io
```
Did it work ok for you?

**Comment by ruslo** [comment]
> Did it work ok for you?

Yes, on several platforms.

What version of CMake you have?

**Comment by pretyman** [comment]
> What version of CMake you have?

3.2.3 on that machine, but must have been a network issue, I tried several hours later and the download was successful.

Thank for the update, I see the `<root-id>/Install/qml/QtMultimedia` now with its contents! :-)

**Comment by ruslo** [comment]
Just finished windows build (visual studio, mingw). It works fine so I've removed `qt-dev` branch and merge it to `develop`.

@pretyman I'm closing this since Qt work all platforms I've tested: Visual Studio, MinGW, Linux, OS X, iOS, Android. Please open new issue if you'll have any problems in future.

**Comment by ruslo** [comment]
> architectures not controlled by toolchain, e.g. it always arm64 + armv7 for device

Link: https://github.com/ruslo/hunter/issues/247

**Comment by headupinclouds** [comment]
> hunter_add_package(Qt COMPONENTS qtmultimedia qtdeclarative)

I'm trying something similar, but am getting errors with qtdelcarative.  I'm curious if it is related.

The following dies for me (android and libcxx toolchains tested) with the latest hunter on the Qt5Declarative include.
```
if(BUILD_QT)
  hunter_add_package(Qt COMPONENTS qtbase qtmultimedia qtdeclarative)
  find_package(Qt5Widgets REQUIRED)
  find_package(Qt5Multimedia REQUIRED)
  find_package(Qt5Declarative REQUIRED)
endif()
```

Results in:
```
CMake Error at CMakeLists.txt:60 (find_package):
  By not providing "FindQt5Declarative.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "Qt5Declarative", but CMake did not find one.

  Could not find a package configuration file provided by "Qt5Declarative"
  with any of the following names:

    Qt5DeclarativeConfig.cmake
    qt5declarative-config.cmake

  Add the installation prefix of "Qt5Declarative" to CMAKE_PREFIX_PATH or set
  "Qt5Declarative_DIR" to a directory containing one of the above files.  If
  "Qt5Declarative" provides a separate development package or SDK, be sure it
  has been installed.
```
@pretyman Were you able to pull in the qt5declarative lib in any project?

**Comment by pretyman** [comment]
Hi,

For the declarative module use `find_package(Qt5Qml REQUIRED)`, however there are some dependencies missing for linking. Unfortunately my machine broke nearly 2 months ago and I couldn't generate Hunter patches to correct the dependencies. I had problems linking in MacOS, iOS and in Linux with the declarative module.

If you get around to working out the dependencies, you can use the `Qt5Widgets_HunterPlugin.cmake` tecnique that @ruslo used. 

Be warned that he put depencies of QtCore together with the Qt5Widgets together into his plug-in, which is not a problem if you are using Qt5Widgets. Since I was not using Qt5Widgets, only declarative, the linking dependencies of Qt5Core didn't come.
